business-networks-workflows/src/main/kotlin/net/corda/bn/flows/DeclineMembershipAttributeChangeFlow.kt

```diff
@@ -0,0 +1,91 @@
+package net.corda.bn.flows
+
+import co.paralleluniverse.fibers.Suspendable
+import net.corda.bn.contracts.ChangeRequestContract
+import net.corda.bn.states.ChangeRequestState
+import net.corda.bn.states.ChangeRequestStatus
+import net.corda.core.contracts.UniqueIdentifier
+import net.corda.core.flows.FlowException
+import net.corda.core.flows.InitiatingFlow
+import net.corda.core.flows.StartableByRPC
+import net.corda.core.flows.InitiatedBy
+import net.corda.core.flows.FlowSession
+import net.corda.core.identity.Party
+import net.corda.core.internal.uncheckedCast
+import net.corda.core.transactions.SignedTransaction
+import net.corda.core.transactions.TransactionBuilder
+
+/**
+ * This flow is initiated by an authorised member who wants to decline a [MembershipModificationRequest] to change
+ * a members attributes.
+ * Modifies an existing [ChangeRequestState]'s status to DECLINED on initiator's and authorised member's ledgers.
+ *
+ * @property requestId The ID of the request which needs to be rejected.
+ * @property notary Identity of the notary to be used for transactions notarisation. If not specified, first one from the whitelist will be used.
+ * */
+@InitiatingFlow
+@StartableByRPC
+class DeclineMembershipAttributeChangeFlow(
+        private val requestId: UniqueIdentifier,
+        private val notary: Party? = null
+) : MembershipManagementFlow<SignedTransaction>() {
+    @Suspendable
+    override fun call(): SignedTransaction {
+        val bnService = serviceHub.cordaService(BNService::class.java)
+
+        val membershipChangeRequest = bnService.getMembershipChangeRequest(requestId)
+                ?: throw MembershipChangeRequestNotFoundException("Could not find change request state with $requestId request ID")
+
+        val membershipId = membershipChangeRequest.state.data.membershipId
+
+        auditLogger.info("$ourIdentity started rejecting membership attribute changes of " +
+                "member with $membershipId membership ID for request with $requestId request ID")
+
+        bnService.getMembership(membershipId)
+                ?: throw MembershipNotFoundException("Membership state with $membershipId linear ID doesn't exist")
+
+        val membershipChangeData = membershipChangeRequest.state.data
+
+        // check whether party had permissions to modify the attributes
+        val networkId = bnService.getMembership(membershipChangeData.membershipId)!!.state.data.networkId
+
+        if (membershipChangeData.pendingBusinessIdentityChange != null) {
```

Comment:
I understand why you're checking for each permission individually, but I'm thinking we can say that authorisation requires both at the same time, regardless of what changes are being proposed?

Reply:
I don't agree with you here, sure it would be easier to check for the permissions that way, but if an authorised party only has role modification permission I think it should be able to at least approve those requests which only asks for role changes.